### ISSUE TITLE
enable querying of old data with new UTF-8 names

### ISSUE THREAD

**Issue by ywwg** [prometheus/prometheus#13765]

**Comment by ywwg** [comment]
@dimitarvdimitrov says:
I managed to go through the doc and am leaving my notes here. I think my biggest questions are around how promQL will handle the mixed representations:

* `-promql.utf8_broad_lookup.until=<date-time>` the doc wasn’t explicit, but it will make more sense if this is based on the samples’ time range that a block covers, not on the block’s ULID. This allows for recompaction of old blocks without breaking the compatibility
* another thing I didn’t see mentioned is changes in the promQL engine. While some series will be queried in their old, others as new and both will be retrieved from the block, how do you want to treat those series in query execution - is `{"label.name"="123", __name__="up"} `joined with `{"label_name"="123", __name__="up"} `in promQL joins or will those two be considered two distinct series by the engine? In other words when I query for {"label.name"="123"} do I get back two series with one sample each or one series with two samples?
  * I think the former makes more sense from a user’s perspective
  * if the latter, then doesn’t this trigger the “multiple vectors with the same label set”
  * as an extention to this - what happens if the query aggregates by a UTF8 label name? `sum by ("label.name") ({"label.name"="123"} ` and `sum by (label_name) ({"label.name"="123"}`
* I don’t fully understand the direction of the support. Building on the example above:
 * `{label_name="123"}` would include both series or only the legacy-style series and return one series with one sample?
 * is there a difference if I query {"label_name"="123"}?

**Comment by ywwg** [comment]
me:

* if the query is for "label.name", series with "label_name" should be considered that same, not distinct series. (This is why we have the caution not to create new utf8 metric or label names that, under substitution, would collide with existing distinct labels -- there's no way to differentiate that case)
* I would call the aggregation query malformed to contain both utf8 and nonutf8 representations. The intent is that the user would only write the utf8 version. If the user did write the query as you did, the `sum by (label_name)` would (should? might? undefined behavior?) not pick up `label.name`  so the results would be incorrect.
* queries for label_name would not correctly return  data for `label.name` . I cannot think of a reasonable way to up-convert such queries. (since underscores are completely ambiguous)

**Comment by ywwg** [comment]
out of scope for v3 but will be a follow-on

**Comment by npazosmendez** [comment]
I've started looking into this and attempting to hack something together. A few questions/thoughts:

#### About semantics

1. Minor nit: the [proposal](https://github.com/prometheus/proposals/blob/main/proposals/2023-11-13-utf8-migration.md#mixed-format-scenarios) suggests flags `-promql.utf8_broad_lookup.*`, but this feels more like a TSDB config rather than a query language config. Maybe rename to `-tsdb.utf8_broad_lookup.*`?

1. Consider this case where we have these two series that we might want to "combine":

    - `{__name__="foo_bar", baz_qux="value"}: (t=1, 1.0)`
    - `{__name__="foo.bar", "baz.qux"="value"}: (t=2, 2.0)`

    Now say the user queries for `{"foo.bar"}`. We want to "convert" `foo_bar` into `foo.bar`, but what about `baz_qux` into `baz.qux`? i.e. should we return a or b?

    a. `{__name__="foo.bar", "baz.qux"="value"}: (t=1, 1.0), (t=2, 2.0)`

    b. `{__name__="foo.bar", "baz_qux"="value"}: (t=1, 1.0)` and `{__name__="foo.bar", "baz.qux"="value"}: (t=2, 2.0)`

    Looking at this isolated example, one might be tempted to say a, but what if:

    - `{__name__="foo_bar", baz_qux="value"}: (t=1, 1.0)`
    - `{__name__="foo.bar", "baz.qux"="value", "baz&qux"="value"}: (t=2, 2.0)`

    What now? Should `baz_qux` become `baz.qux` or `baz&qux`? Or both? Or neither?

    This is related to collisions, which the [proposal](https://github.com/prometheus/proposals/blob/main/proposals/2023-11-13-utf8-migration.md#non-goals) specifically says won't/can't be handled. So there is no _correct_ answer. I thought of a few options, in increasing order of "coverage":

    
    1. Only the `__name__` label is relabelled, and only in the presence of a `__name__="bla"` equality matcher. In this case, the answer would be "b".
        - Con: won't help with migrations where a label name changes (e.g. `foo_bar=value` to `foo.bar=value`). For example, the query `sum by ("baz.qux") ({"foo.bar"})` would not relabel `baz_qux` to `baz.qux`.
    1. Only the labels present in equality matchers are relabelled. In this case, the answer would be "b" for the first example, but would be "a" if the query was `{"foo.bar", "baz.qux"="value"}`.
        - Challenge: what if the query was `{"foo.bar", "baz.qux"="value", "baz&qux"="value"}`?
        - Con: complex semantics for the case above.
    1. Only the labels present in equality matchers or aggregations are relabelled. E.g. `sum by ("baz.qux") ({"foo.bar"})` would relabel `baz_qux` to `baz.qux` and `foo_bar` to `foo.bar`.
        - Challenge: same as above, what if the query was `sum by ("baz.qux", "baz&qux") ({"foo.bar"})`?
        - Con: complex semantics for the case above.
        - Con: implementation-wise, this likely impacts both the query engine and the TSDB, and passing context between the two.
    1. Relabel all labels found in the resulting series by trying all possible combinations. Simplified algorithm: when querying, take all non-escaped labels found in a UTF8 block, escape them, and see if they are present in the non-UTF8 blocks - if present, then relabel to the label in UTF8.
        - Challenge: similar as above, what if we had `{"foo.bar", "baz.qux"="value", "baz&qux"="value"}`?
        - Challenge: what if we don't have a UTF8 block to take as the source of truth?
        - Con: even more complex semantics for the cases above.
        - Con: not an expert in the tsdb, but this sounds overly complex to implement, as one would need to first look inside UTF8 blocks, and use that context to relabel non-UTF8 blocks.

    
    Right now, I'm leaning towards picking the simplest approach, both for implementation and for user understanding. That would be option 1. I also have a feeling that it will be _mostly_ metric names that will be affected in a migration, but that's just a guess. Also, we can start with the simplest approach and expand later if needed.

#### About implementation


1. I see two ways of implementing the query explosion in a block:
    1. Tweak the `PostingsForMatchers` so it transparently returns both results, and `blockQuerier.Select` so the series are combined and relabelled correctly.
    2. Add a new `Querier` implementation (e.g. `escapedUTF8BlockQuerier`) that wraps the `blockQuerier` with the following logic: escape the matchers and relabel the results. The final result would be the union of the results of the `blockQuerier` and the `escapedUTF8BlockQuerier`.

    I like option 2, because it's the less invasive of the two. It doesn't feel right to complicate and interfere with the core logic of the tsdb. There are two caveats though:
    - It could be considered weird that a block has metadata that is not obeyed by the `blockQuerier`, but only by the `escapedUTF8BlockQuerier` wrapper. I think this is minor though.
    - Didn't get that far yet, but I suspect it will be trickier to optimize (compared to option 1).


1. As @dimitarvdimitrov said, we must ensure we return the relabelled results ordered by their new labels 
(*). Now, to do this, I don't see a way around having to first load all the series' labels in memory, relabel them, and then sort them. I'm not sure how much of a performance hit this will be. I would like some feedback on this from someone more experienced with the tsdb. 




(*) BTW, I don't _fully_ understand why this is a requirement. Is the merging of various results the only reason? Or is there also an enforcement in the spec about how results are returned?

**Comment by npazosmendez** [comment]
Just had a realization: if the query explosion + relabeling is scoped to `__name__` and for cases where we have a  `__name__="foo.bar"`  matcher, then the ordering of the series does not change after relabeling, because:
- The ordering of the labels sets does not change, since no label names change
- Since we have a `__name__="foo_bar"` matcher, we are guaranteed that all series will have the `__name__="foo_bar"`  label. If we change them to `__name__="foo.bar"` for _all_ series, the lexicographical order remains the same, because that symbol comparison is no-op for the sorting.

This is another advantage of option (i) I described above:

> Only the `__name__` label is relabelled, and only in the presence of a `__name__="bla"` equality matcher.

____________________________

Worth noting that if relabeling arbitrary label names, that property does not hold. E.g. say we rename `_a` to `áa`:

- Original label sets (sorted): `{"_a"="1", "b"="2"}` <  `{"_a"="2", "b"="1"}` 
- Renamed label sets (sorted):  `{"b"="2", "áa"="1"}` > `{"b"="1", "áa"="2"}` 

Note: `"_a" < "b" == true` and `"b" < "áa"  == true`

**Comment by dimitarvdimitrov** [comment]
### Sorting

> BTW, I don't fully understand why this is a requirement

The `Querier` interface has an parameter which can optionally force the resulting `SeriesSet` to be sorted. The only reason I can think of that we need them sorted is because of merging of samples&chunks

https://github.com/prometheus/prometheus/blob/d5f68872943062d4535ef8cffc53b4613a541d11/storage/interface.go#L115-L117


> * The ordering of the labels sets does not change, since no label names change

Aren't series sorted by label names, then by label values? For example

```
__name__=foo.bar b=1
__name__=foo_bar a=1
```

once they're relabelled, they sort the other way:

```
__name__=foo_bar a=1
__name__=foo_bar b=1
```

**Comment by npazosmendez** [comment]
Thanks for the input @dimitarvdimitrov.

> ```
> __name__=foo.bar b=1
> __name__=foo_bar a=1
> ```

 Sorry, there's one detail I had in mind but didn't mention.  You are right with what you are saying, but I'm thinking of an algorithm where `__name__=foo.bar` and `__name__=foo_bar` are queried separately, and then merged (option 2 for implementation I mentioned [here](https://github.com/prometheus/prometheus/issues/13765#issuecomment-2346688752:~:text=Add%20a%20new%20Querier%20implementation%20(e.g.%20escapedUTF8BlockQuerier)%20that%20wraps%20the%20blockQuerier%20with%20the%20following%20logic%3A%20escape%20the%20matchers%20and%20relabel%20the%20results.%20The%20final%20result%20would%20be%20the%20union%20of%20the%20results%20of%20the%20blockQuerier%20and%20the%20escapedUTF8BlockQuerier.)). So, except when merging the results from both query variations, you would never have mixed  `__name__=foo.bar` and `__name__=foo_bar` . The renaming would happen in a set where all series have the `__name__=foo_bar` label, and is converted into `__name__=foo.bar`

**Comment by npazosmendez** [comment]
I'll add an option that was mentioned in the sync we had: no relabeling at all; the query is exploded but the series with their labels are returned as they are in the store. If needed, the user would tweak the query to deal with that.

**Comment by beorn7** [comment]
While discussing https://github.com/prometheus/prometheus/issues/13586, I suddenly remembered the implications for this issue.

As we want to strictly require a selector to return only matching results, we _do_ have to do some label rewriting in the result (or at least _adding_ the relevant labels). This could be restricted to the labels explicitly queried for so that there are no ambiguities about how to translate them.

**Comment by beorn7** [comment]
Also, for labels that are involved in label matching (in aggregations, but even for binary operators), we need to do the rewriting in the result anyway, e.g. you have `foo + bar`, and there exists `foo{"f.b"="baz"} 3` in a new block and `bar{f_b="baz"} 2` in an old (or mixed) block. I guess that should mean that the label match succeeds, and we get `{"f.b"="baz"} 5` as the result.

Too bad that this is so complicated.

**Comment by npazosmendez** [comment]
> Also, for labels that are involved in label matching (in aggregations, but even for binary operators), we need to do the rewriting in the result anyway, e.g. you have foo + bar, and there exists foo{"f.b"="baz"} 3 in a new block and bar{f_b="baz"} 2 in an old (or mixed) block. I guess that should mean that the label match succeeds, and we get {"f.b"="baz"} 5 as the result.

How would you know that `f_b` has to be renamed to `f.b` and not `f,b`/`f#b`/`f!b`/etc ? That would require going through all series' labels, escape them and try to match them to something from the other side (which can also have collisions, but that's another story). If I understood you well, that sounds a bit too complex and possibly unperformant to me. 

An alternative to that would be to ask the user to list the labels it wants exploded+renamed as part of the matchers, as Owen mentioned [here](https://github.com/prometheus/prometheus/issues/13765#issuecomment-2356562064): `foo + bar{"f.b"=".*"}`

**Comment by npazosmendez** [comment]
If we _really_ need to do the relabeling (because of  https://github.com/prometheus/prometheus/issues/13586), I would like to bring back one of my [previous ideas](https://github.com/prometheus/prometheus/issues/13765#issuecomment-2346688752): only explode+relabel the `__name__` label value. That simplifies the feature significantly, and it could cover a good number of cases. 

(While typing this again I realized that dots in label names are common in OTel, aren't they? E.g. `kubernetes.namespace="asd"`. Anyway, I think it's still worth considering as a first iteration, it simplifies everything)

**Comment by ywwg** [comment]
yes, otel has a lot of dots in label names



> How would you know that `f_b` has to be renamed to `f.b` and not `f,b`/`f#b`/`f!b`/etc

maybe:

1. perform query as-is  (in this example, `foo + bar`)
2. for every returned label, is it utf8? (we see labels `f.b` and `f_b`)
3. if so, combine with other matching escaped labels (combine to `f.b`)

**Comment by npazosmendez** [comment]
> 3. if so, combine with other matching escaped labels (combine to f.b:5)

I'm not sure what you mean by this, I don't think we want to combine and aggregate values, just rename labels.

In any case, I think doing an all-to-all comparison of labels to try to match a probably-escaped string into a probably-unescaped string sounds too complex, at least for a first iteration of this feature. If we are relabeling any labels, I prefer your original suggestion of asking the user to put a noop matcher `{"f.b"=".*"}`.

Also, I insist with [my idea here](https://github.com/prometheus/prometheus/issues/13765#issuecomment-2379162543).

**Comment by ywwg** [comment]
I am referencing the combining that bjorn mentioned:

> we need to do the rewriting in the result anyway, e.g. you have `foo + bar`, and there exists `foo{"f.b"="baz"} 3` in a new block and `bar{f_b="baz"} 2` in an old (or mixed) block. I guess that should mean that the label match succeeds, and we get `{"f.b"="baz"} 5` as the result.

So I guess, relabel f_b as f.b.  

> all-to-all comparison of labels

It's not n*n, it's just n.  if I get back a label "f.b", I only have to check if there is also "f_b" and relabel if it's there. That's 1 extra check for each single utf-8 label.  If I also get back f_c I don't need to do anything with that unless I also saw "f.c".

**Comment by ywwg** [comment]
(I am not sure if labels are stored in a map or a list -- if it's a map then the lookup is cheap and if it's a list then it's a logn binary search of that list for the matching escaped label)

**Comment by npazosmendez** [comment]
What would you do for a query like `foo{}`, when your store has `foo{"b_r"="baz"}` and `foo{"b.r"="baz"}`? Would you return a single series `foo{"b.r"="baz"}` (merging the results), or the two `foo{"b_r"="baz"}` and `foo{"b.r"="baz"}`?

**Comment by ywwg** [comment]
return one, `foo{"b.r"="baz"}`, under the assumption that because b.r->b_r, the second one is escaped.

**Comment by npazosmendez** [comment]
That sounds a bit too magical to me. Here's one scenario where it can get weird: what if you you are querying a time range for which you don't have the `"b.r"` label name in that series (only `"b_r"`)? You would return different results depending on the time range queried. 

(Also I would expect that to be complex to implement, because you have to combine the results of various blocks with the context of the data you got from other blocks. But not an expert here)

**Comment by ywwg** [comment]
> You would return different results depending on the time range queried.

yes, although if the user were to rewrite the query then that would provide a hint:

`foo{} offset $old_data`

→ `foo{"b_r"}:5`

`foo{"b.r"=~".*"} offset $old_data`

→ `foo{"b.r"}:5`

(we now elide the `.*` but prom could still use the information in the query string to inform the relabeling)


The concrete use-case I'm thinking about is a dashboard with a query like `{"my.api", "http.status"=~"4..."}`.  That will return the expected results no matter the time range.  The only time it would be potentially confusing is Explore queries, but most users don't do that anyway.  

I see the edge cases you're pointing out but I am not convinced they are something that would trip up users. I feel like there are two classes of user:
* the people managing the migration, who have explicitly chosen to enable this feature and will have the documentation for how the behavior works, including edge cases
* other more casual dashboard users.

I think the casual users could be told "always query with the new dots syntax, it should always return the right results."  And then maybe the follow-on is "if you're querying old data, you may still see the old escaping in the UI, but queries can still always use the dots.  I guess there's a UI issue where Grafana might suggest label names with underscores.... But this is starting to feel more and more edge-casey.  If the sophisticated engineer who configured the system thought that was too much, they can choose not to use the migration flags.

I guess my point is, because this is an opt-in feature, we can code a minimal feature with edge-cases and document them as Known Issues, and then we can find out from users if the edge cases are actually a problem or not.

**Comment by ywwg** [comment]
we can also revisit this in a call after I get back from PTO  :)

**Comment by npazosmendez** [comment]
> foo{"b.r"=~".*"} offset $old_data

Ok so you do think we should rely on "b.r"=~".*" hints to know which labels to remap. But I guess you don't think that should be the _only_ way to make the tsdb relabel things, but it should also attempt to guess.


> I guess my point is, because this is an opt-in feature, we can code a minimal feature with edge-cases and document them as Known Issues, and then we can find out from users if the edge cases are actually a problem or not.

I agree with you here (and with the previous paragraphs too), but the feature you are suggesting does not sound minimal to me. Here are the options I picture in my head, roughly, in increasing order of complexity (both in how it works and implementation-wise) and "magic" (both in the good and the bad sense):

1. ~~Explode queries but don't relabel~~ (we can't do that because of the issue Bjorn pointed out)
1. Explode and relabel only the `__name__` matcher, when it's non-legacy UTF8.
1. Explode all matchers that are non-legacy UTF8, relabel only the labels that are mentioned in the matchers
1. Explode all matchers that are non-legacy UTF8, relabel only the labels that are mentioned in the matchers, aggregations or groupings
1. Explode all matchers that are non-legacy UTF8, relabel the labels that are mentioned in the matchers, aggregations or groupings, and also the labels that are the escaped version of another label seen in the results.
2. ... ?


What we are discussing sounds like option 5. 

With the same idea of "let's try something and see how useful it is and iterate", I would start from a simpler version of the feature, and then cover more cases if they really come up as a problem and justify the extra complexity and effort.

**Comment by npazosmendez** [comment]
> we can also revisit this in a call after I get back from PTO :)

Yeah, let's do that :+1:. I'll probably pause any implementations progress until then, but I'll also talk about this with Bjorn next week.

**Comment by beorn7** [comment]
Thanks for all the good arguments and findings here. It occurs to me now that this problem is way bigger than just exploding the queries.

I would even argue that the quely explosion is the part that is easiest to do by the user themself: They could just modify the query to include all the various versions of a name that they want to cover.

However, where some kind of "auto migration" is way more important is where the user doesn't state the names in question at all.

The first thing is _range selectors_: Let's take the query `rate(http.requests_total[5m])`. The metric is called `http_requests_total` in the old world, and the user wants the old and the new metric to belong to the same series so that the rate calculation still works seamlessly. (Even more relevant for very long ranges for monthly SLOs and such.) Maybe the implicit rewriting of the result would even solve that case, if we implemented it properly (but it could be very hard to implement). But that's still a case where the name in question is spelled out. Now imagine this metric also has a label `kubernetes.namespace`. To still make the old and new series the same, we also have to do the explosion and rewriting for a label name that was never mentioned in the query.

The second thing is label matching. If the user writes `a + b`, they want that the same labels match on both sides of the operator. If `a` is old and `b` is new, we again have to implicitly do the matching on the exploded label names (which is something discussed above in good detail). I think it is possible and maybe not even too expensive, but again much more complicated than rewriting a query before submitting it to the PromQL engine, and then rewrite the results. There is an error case where multiple labels on the `b` (`x.y`, `x/y`) side might match with the same label on the `a` side (`x_y`). But I guess we could get away with error'ing out of this (presumably rare) case.

We can think about this again after a few nights of sleep (and with @ywwg back on board), but currently it looks like the following to me:
- To make the migration feature actually useful, we have to do way more than query rewriting.
- We have to apply the "wide" querying to labels that aren't explicitly mentioned in the query.
- We have to apply quite invasive changes somewhere at the border between storage and query engine to make sure that series identity is handled in a useful fashion.

In summary, this is _way_ more complex than I have anticipated. I wouldn't say it's infeasible, but it really looks like a quite involved project that we should only tackle if there is a strong need. It also makes other approaches appear more feasible in comparison, e.g. rewriting storage blocks.

**Comment by ywwg** [comment]
I am not yet going to mark this as wontfix but given the increasing apparent complexity and absent compelling need we have no plans to work on this.
